### PR TITLE
Fix package not found error

### PR DESCRIPTION
**Pull request by jordigg**
After splitting the output returned by the commands run via Brew another error appeared because we'll still parsing the brew cask output looking for error messages.

Tested on fresh machines on 10.11 and 10.12. Brew and brew cask apps get installed successfully.

### CODE REVIEW

**Comment by jordigg** [comment]
@TheKevJames another PR sent your way, sorry again for the hassle... last change with the output split caused all brew cask packages to return "package not found". Fixed now and tested on fresh machines on El Capitan and Sierra.

**Comment by jordigg** [comment]
Test failed for timeout. Not sure why it took that long. @TheKevJames can you check?

**Comment by TheKevJames** [comment]
Good catch, I'd missed this!

This PR seems to cause the opposite issue: now all brew cask packages are marked as correctly installed.

```
Debug: Executing with uid=501 gid=20: '/usr/local/bin/brew cask install gtasmnfhbamehjbfaerjhbvsrehb'
Notice: /Stage[main]/Main/Package[gtasmnfhbamehjbfaerjhbvsrehb]/ensure: created
```

Are we sure that we have enough information from stdout only to get reliable results?

Test failures: looks to be unrelated. I've opened #53 to address that.

**Comment by jordigg** [comment]
When trying to install an non-existing package the exit code is 1 (fail) so puppet should fail at this point. I'm investigating now.
Did check beforehand the option of [failonfail ](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/execution.rb#L46)options but I assumed was set to true by default so not sure why succeeds.

**Comment by TheKevJames** [comment]
I remember running into this when I was first writing this package -- the TLDR is that I had issues getting return codes to propagate correctly, which is why we do the stderr parsing in the first place.

**Comment by jordigg** [comment]
Brew error reporting is just a mess.
Just tested with the `failonfail` flag to true and did work. I get error when package doesn't exist but for example the `brew list --versions` of that package returns 0 and the same but without `--versions` returns 1.
Now I even got an error that didn't show up before when I was trying to tap a repo with the wrong name. It never failed before so I was unaware.
I think we should rely on exit codes rather than trying to parse messages. I will report this "wrong" error codes to the brew core project.

**Comment by TheKevJames** [comment]
@jordigg I don't think we can merge this until the return codes are correct for all of the commands we use (or, at least, we can't release this). I've opened the attached issue on brew, if you discover any other such issues please do the same. I will look at this PR again once upstream is merged.

**Comment by TheKevJames** [comment]
Looks like homebrew fixed the return codes upstream (thanks @homebrew-devs!). I've restarted the tests on this PR.

**Comment by jordigg** [comment]
Still failing? Are the tests correct? We are checking versions of packages which aren't installed. They return 1 (error) which is correct (I guess) since the package is not present. Is that the expected behaviour?

**Comment by jordigg** [comment]
Now I'm having issues with the package `wget`. Right now for the packages I use is the only one that give me problems.
It fails on `brew list --versions wget` when trying to install the package, the others install just fine.
```
Could not evaluate: Could not list packages: Execution of '/usr/local/bin/brew list --versions wget' returned 1:
/opt/puppetlabs/puppet/cache/lib/puppet/provider/package/brew.rb:50:in `rescue in package_list'
```

The error is fine because the package is not installed on the machine but I don't understand why it does that check before trying to install it.

@TheKevJames do you know why that check before a package install?

**Comment by jordigg** [comment]
A quick fix for this is using `execute([command(:brew), :cask, :**info**, name])` which verify there's a package available on brew for that package. Can we cleaned up but if we move away from parsing messages to work with error codes there are many things than can be taken away and this way rely only on brew output rather than implementing similar logic twice.
```
def self.package_list(options={})
    Puppet.debug "Listing installed packages"
    list = []
    begin
      if name = options[:justme]
        result = execute([command(:brew), :cask, :info, name])
      else
        result = execute([command(:brew), :list, '--versions'])
        list = result.lines.map {|line| name_version_split(line)}
      end
      Puppet.debug "Found packages #{result}"
    rescue Puppet::ExecutionFailure => detail
      raise Puppet::Error, "Could not list packages: #{detail}"
    end

    if options[:justme]
      return list.shift
    else
      return list
    end
  end
```

**Comment by TheKevJames** [comment]
We do the check prior to installing the package to see whether the package is installed already (ie. whether we need to install it). It looks like now that we have `failonfail` set, this is causing problems where the pre-install check is failing (which is expected) and throwing an error (which is not).

**Comment by jordigg** [comment]
Yes just saw it. Already fixed it with the code I sent on my last message but I'll clean it a bit. There's no different exit code for when the package is already installed so we will need to parse text in this case.
I think the `brew list package` is to check if something is already installed but doesn't check if there's a package available to be installed. For that you need the `brew info package`.
Will work on that and update the PR tomorrow.

**Comment by TheKevJames** [comment]
I get a different exit code when the package is installed:

```
[kevinc@atomicbox(zsh):~] brew list --versions wget
wget 1.18
[kevinc@atomicbox(zsh):~] echo $?
0
[kevinc@atomicbox(zsh):~] brew list --versions this-package-is-not-installed
[kevinc@atomicbox(zsh):~] echo $?
1
```

This was fixed in the brew PR I linked to -- you may need to `brew update`?

**Comment by jordigg** [comment]
Yes that's correct. failonfail doesn't work on that specific case. I'm now trying with passing the failonfail value on the exec so we can choose if we want it to true or false.
For example, on `brew install package` we want `failonfail = true` but on `brew list --versions package` we need it false.

**Comment by jordigg** [comment]
There are two options after brew fixed that exit code:

1- Go back to `failtofail = false`
and rely on text parsing like we have now on the install to throw error when package doesn't exist.
```
if output =~ /Searching taps/
      raise Puppet::ExecutionFailure, "Could not find package #{name}"
    end
```

2- Continue using `failtofail = true` and raise error on install
```
begin
      Puppet.debug "Installing #{name}"
      output = execute([command(:brew), :install, name, *install_options], true)
    rescue Puppet::ExecutionFailure => detail
      raise Puppet::Error, "Could not find package: #{name}"
    end
```

I've been testing for a while, I would try to move away from text parsing as much as possible. Here is the branch with my latest changes I'm using just to test. When we agree on something I'll complete this PR. This should pass the tests that were failing before.

https://github.com/improbable-io/puppet-homebrew/tree/test

**Comment by TheKevJames** [comment]
I'm a fan of "option 2" here; it would certainly be nice to move us away from text parsing. If you were to open a PR migrating to that method, I'd happily review+accept it.

The demo code in your comment looks good, though just a quick request: please use a named parameter for setting `failonfail`, ie.

    output = execute([command(:brew), :install, name, *install_options], failonfail: true)

Other than that, looks great!

**Comment by jordigg** [comment]
there is one failure: https://travis-ci.org/TheKevJames/puppet-homebrew/jobs/167628983#L8680
It's caused by a bad test. Is trying to install `clementine` from brew when it should be installed from brewcask. Is that expected as part of the test or should it be changed to `brewcask`?
https://github.com/TheKevJames/puppet-homebrew/blob/2bb8c8ad9a6ed08e1b35cad63dd1216697b582ea/tests/packages.pp#L13
```
package { 'clementine':  # in brewcask
  ensure   => present,
  provider => homebrew,
}
```

**Comment by jordigg** [comment]
Oh I see... My bad, will send a fix. Homebrew should check brew first and then brewcask.

**Comment by jordigg** [comment]
There's another weird error code from brew. If the package is not found on brew but then is found and is installed successful on another tap the exit code is still 1. That should be fixed and return 0 if the package gets installed.

**Comment by jordigg** [comment]
All green. I've seen you have removed some debug messages. On my PR I think I have quite a bit of them, do you want them removed?

File: lib/puppet/provider/package/brew.rb [hunk 1]
```diff
@@ -8,13 +8,16 @@
...
-      raise Puppet::ExecutionFailure, "Could not find package #{name}"
+    begin
+      Puppet.debug "Looking for #{name} package..."
+      output = execute([command(:brew), :info, name], failonfail: true)
+    rescue Puppet::ExecutionFailure => detail
+      raise Puppet::Error, "Could not find package: #{name}"
     end
 
+    Puppet.debug "Package found, installing..."
+    output = execute([command(:brew), :install, name, *install_options])
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brew.rb]: Would setting this to `failonfail` be a good idea? We want some way to warn the user if a package installation failed: I'm thinking we wrap this in a `begin-rescue` like the `info` command above and raise a package installation error on fail.

Please also check to make sure the checksum check below counts as success, or move it into the `resuce` block as a potential recovery.

File: lib/puppet/provider/package/brew.rb [hunk 2]
```diff
@@ -68,4 +77,4 @@ def self.name_version_split(line)
       nil
     end
   end
-end
+end
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brew.rb]: newlines at ends of files, please

File: lib/puppet/provider/package/brew.rb [hunk 3]
```diff
@@ -40,9 +43,15 @@ def self.package_list(options={})
...
-        result = execute([command(:brew), :list, '--versions', name])
+        result = execute([command(:brew), :list, '--version', name])
+        if result.empty?
+          Puppet.debug "Package #{result} not installed"
+        else
+          Puppet.debug "Found package #{result}"
+        end
       else
         result = execute([command(:brew), :list, '--versions'])
+        Puppet.debug "Found packages #{result}"
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brew.rb]: remove

File: lib/puppet/provider/package/brew.rb [hunk 4]
```diff
@@ -40,9 +43,15 @@ def self.package_list(options={})
     Puppet.debug "Listing installed packages"
     begin
       if name = options[:justme]
-        result = execute([command(:brew), :list, '--versions', name])
+        result = execute([command(:brew), :list, '--version', name])
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brew.rb]: Why switch to `--version` here?

File: lib/puppet/provider/package/brewcask.rb [hunk 5]
```diff
@@ -8,15 +8,16 @@
...
-      raise Puppet::ExecutionFailure, "Could not find package #{name}"
+    begin
+      Puppet.debug "Looking for #{name} package..."
+      output = execute([command(:brew), :cask, :info, name], failonfail: true)
+    rescue Puppet::ExecutionFailure => detail
+      raise Puppet::Error, "Could not find package: #{name}"
     end
 
+    Puppet.debug "Package found, installing..."
+    output = execute([command(:brew), :cask, :install, name, *install_options])
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brewcask.rb]: See comments above on `failonfail`, `rescue` block, etc

File: lib/puppet/provider/package/brewcask.rb [hunk 6]
```diff
@@ -43,10 +44,15 @@ def self.package_list(options={})
...
       if name = options[:justme]
         result = execute([command(:brew), :cask, :list, '--versions', name])
+        if result.empty?
+          Puppet.debug "Package #{result} not installed"
+        else
+          Puppet.debug "Found package #{result}"
+        end
       else
         result = execute([command(:brew), :list, '--versions'])
+        Puppet.debug "Found packages #{result}"
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brewcask.rb]: remove

File: lib/puppet/provider/package/brewcask.rb [hunk 7]
```diff
@@ -71,4 +77,4 @@ def self.name_version_split(line)
       nil
     end
   end
-end
+end
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brewcask.rb]: newlines at ends of files, please

File: lib/puppet/provider/package/brewcommon.rb [hunk 8]
```diff
@@ -91,4 +91,4 @@ def fix_checksum(files)
     end
     raise Puppet::ExecutionFailure, "Checksum error for package #{name} in files #{files}"
   end
-end
+end
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brewcommon.rb]: newlines

File: lib/puppet/provider/package/homebrew.rb [hunk 9]
```diff
@@ -8,18 +8,20 @@
...
-    if output =~ /Searching taps/
-      Puppet.debug "Falling back to brew-cask (still installing #{name}"
-      output = execute([command(:brew), :cask, :install, name, *install_options])
-      # brewcask includes some funky beer characters that f*ck with encoding
-      output = output.encode('UTF-8', :invalid => :replace, :undef => :replace)
-
+    begin
+      Puppet.debug "Looking for #{name} package on brew..."
+      output = execute([command(:brew), :info, name])
       if output.empty?
```
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: Can we move away from parsing output here?

File: lib/puppet/provider/package/homebrew.rb [hunk 10]
```diff
@@ -8,18 +8,20 @@
...
+         output = execute([command(:brew), :cask, :install, name, *install_options])
+      else
+        Puppet.debug "Package found, installing..."
+        output = execute([command(:brew), :install, name, *install_options], failonfail: true)
       end
+    rescue Puppet::ExecutionFailure => detail
+      raise Puppet::Error, "Could not find package: #{name}"
     end
 
     if output =~ /sha256 checksum/
```
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: See comments above about checking to be sure we don't need to move this into the rescue block.

File: lib/puppet/provider/package/homebrew.rb [hunk 11]
```diff
@@ -52,11 +54,16 @@ def self.package_list(options={})
...
         end
+        if result.empty?
+          Puppet.debug "Package #{result} not installed"
+        else
+          Puppet.debug "Found package #{result}"
+        end
       else
         result = execute([command(:brew), :list, '--versions'])
         result += execute([command(:brew), :cask, :list, '--versions'])
+        Puppet.debug "Found packages #{result}"
```
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: remove

File: lib/puppet/provider/package/homebrew.rb [hunk 12]
```diff
@@ -52,11 +54,16 @@ def self.package_list(options={})
         unless result.include? name
           result += execute([command(:brew), :cask, :list, '--versions', name])
         end
+        if result.empty?
```
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: I don't think we need this all over

File: lib/puppet/provider/package/homebrew.rb [hunk 13]
```diff
@@ -81,4 +88,4 @@ def self.name_version_split(line)
       nil
     end
   end
-end
+end
```
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: newline

File: lib/puppet/provider/package/brew.rb (hunk 1 above)
Comment (jordigg) [inline lib/puppet/provider/package/brew.rb]: Agreed. Will update now.

File: lib/puppet/provider/package/brew.rb (hunk 4 above)
Comment (jordigg) [inline lib/puppet/provider/package/brew.rb]: Typo. Fixed.

File: lib/puppet/provider/package/homebrew.rb (hunk 12 above)
Comment (jordigg) [inline lib/puppet/provider/package/homebrew.rb]: this is just to show debug if the package was found or not. Should we remove this part?

**Comment by jordigg** [comment]
How can I test the `if output =~ /sha256 checksum/` part? I think if we use `failonfail=true` we don't need to parse this errors anymore, we'll just return the detail of the error but not sure how to test it.

File: lib/puppet/provider/package/homebrew.rb (hunk 12 above)
Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: I don't much care one way or the other. Feel free to leave it.

**Comment by TheKevJames** [comment]
Easiest way to reproduce the sha mismatch bug is:

1) install a package
2) uninstall that package
3) manually modify that package's archive file
4) install that package again -- it should hit the sha mismatch block

File: lib/puppet/provider/package/homebrew.rb (hunk 9 above)
Comment (jordigg) [inline lib/puppet/provider/package/homebrew.rb]: Did try with a nested rescue but doesn't work. I can make it continue but after executing everything raises the error anyway. Any ideas how can be done without parsing?

```
 def install
    name = install_name

    begin
      Puppet.debug "Looking for #{name} package on brew..."
      output = execute([command(:brew), :info, name], failonfail: true)
      Puppet.debug "Package found, installing..."
      output = execute([command(:brew), :install, name, *install_options], failonfail: true)
      if output =~ /sha256 checksum/
        Puppet.debug "Fixing checksum error..."
        mismatched = output.match(/Already downloaded: (.*)/).captures
        fix_checksum(mismatched)
      end
    rescue => detail_brew
      next
      begin
        Puppet.debug "Package #{name} not found on Brew. Trying BrewCask..."
        output = execute([command(:brew), :cask, :info, name], failonfail: true)
        Puppet.debug "Package found on brewcask, installing..."
        output = execute([command(:brew), :cask, :install, name, *install_options], failonfail: true)
      rescue => detail_cask
        raise Puppet::Error, "Could not install package: #{detail_cask}"
      end
      raise Puppet::Error, "Could not install package: #{detail_brew}"
    end
  end
```

Comment (TheKevJames) [inline lib/puppet/provider/package/homebrew.rb]: Tried to puzzle through the logic for a few minutes, but the nested rescues turned pretty confusing. Let's leave this as-is for now.

**Comment by jordigg** [comment]
I'll send the requested changes. The part about the checksum check may need more work.
I've tried manually and it's not just an exception. Brew tries to install it anyway building from source. I did try with the `wget`package. It got installed successfully so I got no error and exit code 0. Even though the output contains the the text we try to parse. 
I'm wondering if on the actual version of the module works as expected.
Should we remove it all together since brew installs the package anyway or we should still try to detect that "warning" and cleanup the files?
For now I'll leave it as it is and maybe we can work it out on another PR.

File: lib/puppet/provider/package/brew.rb [hunk 14]
```diff
@@ -8,18 +8,26 @@
...
+
+      if output =~ /sha256 checksum/
+        Puppet.debug "Fixing checksum error..."
+        mismatched = output.match(/Already downloaded: (.*)/).captures
+        fix_checksum(mismatched)
+      end
+    rescue Puppet::ExecutionFailure => detail
+      raise Puppet::Error, "Could not install package: #{detail}"
     end
+
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brew.rb]: extra newline

File: lib/puppet/provider/package/brewcask.rb [hunk 15]
```diff
@@ -1,74 +1,94 @@
-Puppet::Type.type(:package).provide(:brewcask,
-                                    :parent => :brewcommon,
-                                    :source => :brewcommon) do
-  desc "Package management using HomeBrew casks on OS X"
+require 'puppet/provider/package'
 
-  has_feature :install_options
+Puppet::Type.type(:package).provide(:brewcommon,
```
Comment (TheKevJames) [inline lib/puppet/provider/package/brewcask.rb]: This looks like you accidentally copied `brewcommon` into `brewcask.rb`?

**Comment by jordigg** [comment]
Yes I did, sorry about that! Maybe to late for me or that I'm using a Mac to access my work windows machine where I have the code and I had to copy between branches to avoid creating too much mess...

**Comment by TheKevJames** [comment]
Looks good. Will merge assuming tests pass. Thanks a bunch for this!

**Comment by jordigg** [comment]
Thank YOU for helping with the review and maintaining this module, happy to help (with all my mistakes!)

**Pull request merged**